# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/libwebrtcglue/WebrtcGmpVideoCodec.h
# Commit: 709078257c62
# Full Hash: 709078257c628fae054382894de5864b1f5f5c2a
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 3
# Description:
#   Bug 1654112 - Re-wire GMP PluginID between codecs and VideoConduit. r=bwc
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D104724
# ==============================================================================

diff -r 3ca24e761cc3 -r 709078257c62 dom/media/webrtc/libwebrtcglue/WebrtcGmpVideoCodec.h
--- a/dom/media/webrtc/libwebrtcglue/WebrtcGmpVideoCodec.h	Wed Feb 10 10:34:14 2021 +0100
+++ b/dom/media/webrtc/libwebrtcglue/WebrtcGmpVideoCodec.h	Wed Feb 10 15:17:29 2021 +0100
@@ -125,8 +125,6 @@
 
   // Implement sort of WebrtcVideoEncoder interface and support refcounting.
   // (We cannot use |Release|, since that's needed for nsRefPtr)
-  virtual uint64_t PluginID() const = 0;
-
   virtual int32_t InitEncode(const webrtc::VideoCodec* aCodecSettings,
                              int32_t aNumberOfCores,
                              size_t aMaxPayloadSize) = 0;
@@ -145,6 +143,10 @@
 
   virtual int32_t SetRates(uint32_t aNewBitRate, uint32_t aFrameRate) = 0;
 
+  virtual MediaEventSource<uint64_t>* InitPluginEvent() = 0;
+
+  virtual MediaEventSource<uint64_t>* ReleasePluginEvent() = 0;
+
  protected:
   virtual ~RefCountedWebrtcVideoEncoder() = default;
 };
@@ -156,8 +158,6 @@
 
   // Implement VideoEncoder interface, sort of.
   // (We cannot use |Release|, since that's needed for nsRefPtr)
-  uint64_t PluginID() const override { return mCachedPluginId; }
-
   int32_t InitEncode(const webrtc::VideoCodec* aCodecSettings,
                      int32_t aNumberOfCores, size_t aMaxPayloadSize) override;
 
@@ -175,6 +175,14 @@
 
   int32_t SetRates(uint32_t aNewBitRate, uint32_t aFrameRate) override;
 
+  MediaEventSource<uint64_t>* InitPluginEvent() override {
+    return &mInitPluginEvent;
+  }
+
+  MediaEventSource<uint64_t>* ReleasePluginEvent() override {
+    return &mReleasePluginEvent;
+  }
+
   // GMPVideoEncoderCallback virtual functions.
   virtual void Terminated() override;
 
@@ -281,8 +289,11 @@
   // Protects mCallback
   Mutex mCallbackMutex;
   webrtc::EncodedImageCallback* mCallback;
-  uint64_t mCachedPluginId;
+  Maybe<uint64_t> mCachedPluginId;
   const std::string mPCHandle;
+
+  MediaEventProducer<uint64_t> mInitPluginEvent;
+  MediaEventProducer<uint64_t> mReleasePluginEvent;
 };
 
 // Basically a strong ref to a RefCountedWebrtcVideoEncoder, that also
@@ -300,7 +311,13 @@
     RegisterEncodeCompleteCallback(nullptr);
   }
 
-  uint64_t PluginID() const override { return mEncoderImpl->PluginID(); }
+  MediaEventSource<uint64_t>* InitPluginEvent() override {
+    return mEncoderImpl->InitPluginEvent();
+  }
+
+  MediaEventSource<uint64_t>* ReleasePluginEvent() override {
+    return mEncoderImpl->ReleasePluginEvent();
+  }
 
   int32_t InitEncode(const webrtc::VideoCodec* aCodecSettings,
                      const WebrtcVideoEncoder::Settings& aSettings) override {
@@ -337,8 +354,6 @@
 
   // Implement VideoEncoder interface, sort of.
   // (We cannot use |Release|, since that's needed for nsRefPtr)
-  virtual uint64_t PluginID() const { return mCachedPluginId; }
-
   virtual int32_t InitDecode(const webrtc::VideoCodec* aCodecSettings,
                              int32_t aNumberOfCores);
   virtual int32_t Decode(const webrtc::EncodedImage& aInputImage,
@@ -348,6 +363,12 @@
 
   virtual int32_t ReleaseGmp();
 
+  MediaEventSource<uint64_t>* InitPluginEvent() { return &mInitPluginEvent; }
+
+  MediaEventSource<uint64_t>* ReleasePluginEvent() {
+    return &mReleasePluginEvent;
+  }
+
   // GMPVideoDecoderCallbackProxy
   virtual void Terminated() override;
 
@@ -415,9 +436,12 @@
   // Protects mCallback
   Mutex mCallbackMutex;
   webrtc::DecodedImageCallback* mCallback;
-  Atomic<uint64_t> mCachedPluginId;
+  Maybe<uint64_t> mCachedPluginId;
   Atomic<GMPErr, ReleaseAcquire> mDecoderStatus;
   const std::string mPCHandle;
+
+  MediaEventProducer<uint64_t> mInitPluginEvent;
+  MediaEventProducer<uint64_t> mReleasePluginEvent;
 };
 
 // Basically a strong ref to a WebrtcGmpVideoDecoder, that also translates
@@ -434,7 +458,13 @@
     RegisterDecodeCompleteCallback(nullptr);
   }
 
-  uint64_t PluginID() const override { return mDecoderImpl->PluginID(); }
+  MediaEventSource<uint64_t>* InitPluginEvent() override {
+    return mDecoderImpl->InitPluginEvent();
+  }
+
+  MediaEventSource<uint64_t>* ReleasePluginEvent() override {
+    return mDecoderImpl->ReleasePluginEvent();
+  }
 
   int32_t InitDecode(const webrtc::VideoCodec* aCodecSettings,
                      int32_t aNumberOfCores) override {